File: server/src/main/java/com/yawa/server/config/monitoring/MetricRequestTagsProvider.java
package com.yawa.server.config.monitoring;

import com.yawa.server.config.mvc.OperationNameProvider;
import com.yawa.server.constants.MetricTags;
import io.micrometer.core.instrument.Tag;
import io.micrometer.core.instrument.Tags;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.actuate.metrics.web.servlet.WebMvcTagsContributor;
import org.springframework.stereotype.Component;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

@Component
public class MetricRequestTagsProvider implements WebMvcTagsContributor {

    @Autowired
    private OperationNameProvider operationNameProvider;

    @Override
    public Iterable<Tag> getTags(HttpServletRequest request,
                                 HttpServletResponse response, Object handler, Throwable exception) {
        return Tags.empty()
                .and(MetricTags.OPERATION,
                        operationNameProvider.getOperationName(request.getMethod(), request.getRequestURI()));
                /*.and("requestId", StringUtils.isNotBlank(response.getHeader(HttpHeaders.REQUEST_ID)) ?
                        response.getHeader(HttpHeaders.REQUEST_ID) : "none-" + UUID.randomUUID());*/
    }

    @Override
    public Iterable<Tag> getLongRequestTags(HttpServletRequest request, Object handler) {
        return null;
    }
}

File: server/src/main/kotlin/com/yawa/server/monitoring/MetricRequestTagsProvider.kt
package com.yawa.server.monitoring

import com.yawa.server.constants.MetricTags
import com.yawa.server.utils.OperationNameProvider
import io.micrometer.common.KeyValue
import io.micrometer.common.KeyValues
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.server.observation.DefaultServerRequestObservationConvention
import org.springframework.http.server.observation.ServerRequestObservationContext
import org.springframework.stereotype.Component

@Component
class MetricRequestTagsProvider(
    @Autowired val operationNameProvider: OperationNameProvider,
) : DefaultServerRequestObservationConvention() {

    override fun getLowCardinalityKeyValues(context: ServerRequestObservationContext): KeyValues {
        return super.getLowCardinalityKeyValues(context).and(additionalTags(context))
    }

    protected fun additionalTags(context: ServerRequestObservationContext): KeyValues {
        var tags = KeyValues.empty()

        val request = context.carrier

        request.userPrincipal?.let {
            tags = tags.and(KeyValue.of(MetricTags.PRINCIPAL, it.name))
        }

        context.carrier.getHeader("user-agent")?.let {
            tags = tags.and(KeyValue.of(MetricTags.USER_AGENT, it))
        }

        operationNameProvider.getOperationName(request.method!!, request.requestURI!!).let {
            tags = tags.and(KeyValue.of(MetricTags.OPERATION, it))
        }

        return tags
    }
}
